docs/idiomatic-maya-guide.md

```diff
@@ -0,0 +1,418 @@
+## Introduction
+Following are some of the explanations thats comes up when we search for the term "idiomatic". They are:
+1. using, containing, or denoting expressions that are natural to a native speaker
+2. appropriate to the style of art or music associated with a particular period, individual, or group.
+
+I do not remember hearing anything about being idiomatic for other languages. I guess it has become synonymous with the advent of golang. Being idiomatic definitely means more than just good coding practices. It tries to provide a common ground for each member of the team to communicate in a way that is understood clearly. Obviously this is not solved by programming language, not even a high level programming language. IMO current day programming languages provide a bunch of dialects and lets its users (i.e. developers) choose the one they like the most. The fact that there is a choice, it hits the team hard later in the project's release lifecycle. The project code may not look as bad as sphagetti but no good either to understand fast and hence implement features, fixes, etc. faster.
+
+## If Maya is Go then why not Idiomatic Go
+Being idiomatic in Maya includes all the idiomatic pieces in Go and also takes into consideration writing code that understands and responds to Kubernetes a lot better. In addition, team at Maya has tried to put all their storage learnings so far into what is being termed as Idiomatic Maya where Maya is control plane for OpenEBS.
+
+
+## Devil is in the details
+Below are some sample code reviews that describes with examples to write idiomatic maya code.
+
+### Naming - Thumb Rule
+- Use names based on what the logic provides and not based on what the logic contains
+
+```go
+import (
+  // alias `algorithm` conveys about what it contains i.e. some algorithm
+  algorithm "github.com/openebs/maya/pkg/algorithm/nodeSelect/v1alpha1"
+
+  // vs.
+
+  // `nodeselect` tries to convey what the logic provides
+  // this seems more natural way to express
+  nodeselect "github.com/openebs/maya/pkg/algorithm/nodeSelect/v1alpha1"
```

Comment:
nit: Also document the `v1alpha1` versioning of code interfaces and add real world code examples to `kubernetes/kubernetes`. Eg. For more examples see `pkg/controller/cronjob` or something else as a time it can be having two sets of API, `v1alpha1` or `v2`. 

This is a good pattern from Kubernetes codebase.